test/e2e/inspection_test.go

```diff
@@ -145,6 +146,84 @@ var _ = Describe("Inspection", func() {
 		}, e2eConfig.GetIntervals(specName, "wait-available")...)
 	})
 
+	It("should re-inspect the annotated BMH", func() {
+		By("creating a secret with BMH credentials")
+		bmcCredentials := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "bmc-credentials",
+				Namespace: namespace.Name,
+			},
+			StringData: map[string]string{
+				"username": bmcUser,
+				"password": bmcPassword,
+			},
+		}
+		err := clusterProxy.GetClient().Create(ctx, &bmcCredentials)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("creating a BMH")
+		bmh := metal3api.BareMetalHost{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      specName + "-inspect",
+				Namespace: namespace.Name,
+			},
+			Spec: metal3api.BareMetalHostSpec{
+				BMC: metal3api.BMCDetails{
+					Address:         bmcAddress,
+					CredentialsName: "bmc-credentials",
+				},
+				BootMode:       metal3api.Legacy,
+				BootMACAddress: bootMacAddress,
+			},
+		}
+		err = clusterProxy.GetClient().Create(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for the BMH to be in registering state")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateRegistering,
+		}, e2eConfig.GetIntervals(specName, "wait-registering")...)
+
+		By("waiting for the BMH to be in inspecting state")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateInspecting,
+		}, e2eConfig.GetIntervals(specName, "wait-inspecting")...)
+
+		By("waiting for the BMH to become available")
+		WaitForBmhInProvisioningState(ctx, WaitForBmhInProvisioningStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateAvailable,
+		}, e2eConfig.GetIntervals(specName, "wait-available")...)
```

Comment:
Similar to the [discussion](https://github.com/metal3-io/baremetal-operator/pull/1382#discussion_r1350129525) in #1382 this has the issue that we're only proving the state-machine path, not that inspection actually worked (e.g `bmh.Status.HardwareDetails` was updated).

As a first step I wonder if we could add a common assertion that `bmh.Status.HardwareDetails` is not empty after each inspection?

Ideally we'd change something about the environment e.g add some device, reinspect then prove it exists in the re-inspected HardwareDetails, but I appreciate that may be difficult and goes beyond the scope of adding a new test-case.

/cc @lentzi90 and @dtantsur for additional input on this suggestion.

Reply:
Good point. What do you think about using external inspection to skip inspection initially? Then we can provide the input and make sure that it is a bit different from what an inspection would result it. That way we get a clear diff to check when we trigger the inspection using the annotation :slightly_smiling_face: 
Bonus: We save some time by not waiting for the first inspection :hourglass:

Reply:
in e2e we have minkube and VMs, I am not sure how we can change VM from our e2e test, if we can do that we replicate the ideal scenario that @hardys has explained.

Reply:
Sorry, I'm confusing you. I mean do an [external inspection](https://github.com/metal3-io/baremetal-operator/blob/main/test/e2e/external_inspection_test.go) where we can actually put any data we want. In the link you can see the hardware details we use for the external inspection test. We can use the same here but make some changes to it first (it is just a string).
Then when we trigger the inspection with the annotation, IPA will find out what the actual hardware details are and correct them. So we can check that there is a diff before and after the inspection.

Reply:
@lentzi90 that could work (and such a scenario was actually [discussed](https://github.com/metal3-io/baremetal-operator/pull/796/files#r586241286) when I implemented the hardwaredetails annotation)

My only concern with that is it's not actually testing re-inspection, instead it's testing moving from inspection-disabled -> inspection (which is also a valid but probably less common scenario for users?) - this means we won't catch e.g any issues with Ironic that could arise from re-inspecting an already inspected host.

As an alternative we could assert HardwareDetails is correct after the first inspect, then keep track of the `operationHistory` timestamps, but unless we forcibly modify the Status.HardwareDetails between inspection runs it's going to be difficult to validate the re-inspection data (I'm assuming we can't modify the VM configuration from the test).

Reply:
Hmm interesting perspective. I'm not familiar with how different the external inspection and normal inspection are treated internally. Are they very different? My impression was that the sequence is more like inspection-disabled -> external inspection -> BMH is now in available state (inspection is not disabled at this point) -> inspect "again" -> BMH is again in available state.
Is there really a difference in the externally inspected and the "normally" inspected states?

Reply:
@lentzi90 the code-path through the BMO is completely different in the external/disabled case:

- We consume the hardwaredetails when inspection is disabled in the [reconcile loop](https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/baremetalhost_controller.go#L146)
- This also means we short-circuit the Inspecting state to [do nothing](https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/baremetalhost_controller.go#L936) (so we don't drive Ironic through any of the normal inspection process)

In this flow the BMH reaches available state and inspection is still disabled (only the hardwaredetails annotation is [removed on consumption I](https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/baremetalhost_controller.go#L283)IRC)

Reply:
Ok that part I get, what I don't understand is how it in any way would affect the actual inspection that is then triggered.
We *do* check the normal inspection flow in a separate test, so I'm here only interested in checking the functionality of the inspection annotation.

Is the "second" inspection taking into account the existing state in any way? I would expect an externally inspected BMH to be indistinguishable from a normally inspected BMH and that they behave in exactly the same way in every situation.

The annotation that disables inspection is removed/overwritten with the annotation that triggers the inspection, this is a good point. Perhaps we should remove it first and ensure that no inspection is started just from that. Then we can add the annotation a bit later to trigger the inspection.

Perhaps the title of the PR is what is causing this confusion. My intention with https://github.com/metal3-io/baremetal-operator/issues/1369 was that we should test the functionality described here: https://book.metal3.io/bmo/inspect_annotation
I don't think the it is relevant in this case how the BMH ended up in available state with some (potentially) outdated hardware details. The only important part is that adding an annotation to it will trigger an inspection that causes the hardware details to be updated.

Reply:
@lentzi90 the state in Ironic will be different if we use external/disabled vs "real" inspection initially, but perhaps that doesn't matter if the main aim is to test the annotation (and we have good confidence that Ironic behaves correctly when inspection is triggered twice since that's the likely scenario in the "hardware changed" use-case described for the annotation)

Reply:
Got it, thanks for explaining this!
It sounds to me like we would be fine with an external inspection for now, and this should allow a pretty simple test where the hardware details actually change. In the CAPM3 e2e tests I know we test re-inspection properly, and perhaps we should add it here also at some point also. I would be happy to hear about use-cases or more reasons for testing this!

Reply:
I have changed the test @hardys @lentzi90 please take a look and let me know if it looks reasonable. Now first we do an external inspection with the wrong hostname and after that, we remove hardware-utils annotation and change `inspect.metal3.io: disable` to `inspect.metal3.io: ""`, bmh is reinspected after that and we can see the right host-name after re-inspection.